Fix shared use of collectors, fixes #18

jansorg opened the pull request:
- This PR fixes usage of shared gauges, counters and histograms.
- The output at '/metrics' looks good on my end. 
- Tests are not passing due to timeouts, but I think that this isn't related to the changed of this PR
- I'd be very happy if this could be merged. If you need corrections, let me know.
- The remaining direct use of the registry, e.g. `io.vertx.ext.prometheus.VertxPrometheusMetrics.TimerPrometheusMetrics`, could be replaced with the factories. But I didn't want to spend too much time before I know if this PR will be accepted.

Previously gauges, counters and histograms shared their name but were
registered more than once. The register calls failed in prometheus' simpleclient
but these errors were silently ignored. The result was that
only the first collector was used for a name.
Metrics like HTTPServerPrometheusMetrics are instantiated multiple times
and thus need to share gauges, counters and histograms to work properly
across different instances of HTTPServers, HTTPClients etc.

This commit adds factories for shared gauges, counters and histograms.
These are created in VertxPrometheusMetrics and passed around to the different child metrics.
The factories are closed in VertxPrometheusMetrics::close(). They can not
be closed earlier as the collectors need to be registered until the last
client (HTTPServerMetrics, ...) is released. The only safe point to to this
is the initial VertxPrometheusMetrics.

Closes #18

jansorg committed:
Fixed usage of shared gauges, counters and histograms.
Previously gauges, counters and histograms shared their name but were
registered more than once. The register calls failed in prometheus' simpleclient
but these errors were silently ignored. The result was that
only the first collector was used for a name.
Metrics like HTTPServerPrometheusMetrics are instantiated multiple times
and thus need to share gauges, counters and histograms to work properly
across different instances of HTTPServers, HTTPClients etc.

This commit adds factories for shared gauges, counters and histograms.
These are created in VertxPrometheusMetrics and passed around to the different child metrics.
The factories are closed in VertxPrometheusMetrics::close(). They can not
be closed earlier as the collectors need to be registered until the last
client (HTTPServerMetrics, ...) is released. The only safe point to to this
is the initial VertxPrometheusMetrics.

jansorg committed:
Updated Vertx to 3.5.1 and the prometheus libraries to the current versions.

nolequen commented:
Hi @jansorg,

Thanks very much for the PR! The changes look good to me. Would you mind to add a test for multiple http server/client metrics creation?

jansorg committed:
Register newly created collectors in registry

jansorg committed:
Add test case

jansorg commented:
@nolequen Thanks for the feedback. I've added a simple test case to test 2 servers and 2 clients. I also fixed the registration of the new collectors, which was broken.

jansorg commented:
I'm not sure  why DisableTest  is failing, tests are passing here. Maybe it's a load issue on Travis causing timeouts, but I'm not sure...

nolequen commented:
@jansorg thanks again!
